onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention_impl.cu

```diff
@@ -0,0 +1,493 @@
+
+      if (h < k_offset) {
+        q[n * S * H_QK + h] = T{};
+      } else if (h < v_offset) {
+        k[n * S * H_QK + (h - k_offset)] = T{};
+      } else {
+        v[n * S * H_V + (h - v_offset)] = T{};
+      }
+    }
+  }
+}
+
+// Grid: (T)
+// Block: 256
+// For memory efficient fMHA from CUTLASS.
+//     Input: TxNx3xH
+//     Output: 3xTxNxH
+// T is token_count
+// B is batch_size
+// S is sequence_length
+// N is num_heads
+// H is head_size
+template <typename T>
+__global__ void TransposeQKV_TN3H_3TNH(
+    const T* input,
+    int32_t N,
+    int32_t H_QK,
+    int32_t H_V,
+    T* q,
+    T* k,
+    T* v,
+    int32_t token_count) {
+  int token_idx = blockIdx.x;
+
+  input += token_idx * N * (H_QK + H_QK + H_V);
+  q += token_idx * N * H_QK;
+  k += token_idx * N * H_QK;
+  v += token_idx * N * H_QK;
+
+  if (token_idx < token_count) {
+    for (int i = threadIdx.x; i < N * H_QK; i += blockDim.x) {
+      int n = i / H_QK;
+      int h = i % H_QK;
+      q[i] = input[n * (H_QK + H_QK + H_V) + h];
+      k[i] = input[n * (H_QK + H_QK + H_V) + H_QK + h];
+    }
+
+    for (int i = threadIdx.x; i < N * H_V; i += blockDim.x) {
+      int n = i / H_V;
+      int h = i % H_V;
+      v[i] = input[n * (H_QK + H_QK + H_V) + H_QK + H_QK + h];
+    }
+  }
+}
+
+template <typename T>
+void InvokeTranspose(
+    const T* query, const T* key, const T* value, T* output,
+    const int batch_size, const int sequence_length,
+    const int num_heads, const int qk_head_size, const int v_head_size,
+    AttentionQkvFormat source_format, AttentionQkvFormat target_format,
+    const int32_t* token_offset, int32_t token_count,
+    cudaStream_t stream) {
+  if (key != nullptr || value != nullptr) {
+    ORT_ENFORCE(source_format == AttentionQkvFormat::Q_K_V_TNH);
+    ORT_NOT_IMPLEMENTED("Only support packed QKV.");
+  } else {
+    ORT_ENFORCE(source_format == AttentionQkvFormat::QKV_TN3H);
+    if (target_format == AttentionQkvFormat::Q_K_V_BNSH) {
+      const dim3 grid(sequence_length, batch_size);
+      TransposeQKV_TN3H_3BNSH<T><<<grid, kMAX_THREADS_PER_BLOCK, 0, stream>>>(
+          query,
+          num_heads,
+          qk_head_size,
+          v_head_size,
+          output,
+          output + batch_size * sequence_length * num_heads * qk_head_size,
+          output + 2 * batch_size * sequence_length * num_heads * qk_head_size,
+          token_offset,
+          token_count);
+    } else if (target_format == AttentionQkvFormat::Q_K_V_TNH) {
+      const dim3 grid(token_count);
+      TransposeQKV_TN3H_3TNH<T><<<grid, kMAX_THREADS_PER_BLOCK, 0, stream>>>(
+          query,
+          num_heads,
+          qk_head_size,
+          v_head_size,
+          output,
+          output + token_count * num_heads * qk_head_size,
+          output + 2 * token_count * num_heads * qk_head_size,
+          token_count);
+    } else {
+      ORT_ENFORCE(target_format == AttentionQkvFormat::QKV_TN3H);
+    }
+  }
+}
+
+template <typename T>
```

Comment:
maybe it's better to move to https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cuda/bert/utils.cuh

Reply:
will refactor later